File: components/rtps_embedded/example/host_pubsub.py
"""
PC-side pub/sub counterpart for the rtps_embedded example firmware.

The firmware runs on an ESP32-Ethernet-Kit acting as a DHCP server at
192.168.4.1.  Connect the host PC to the same Ethernet segment so it
obtains an address in the 192.168.4.0/24 subnet.

Roles mirror the firmware:
  initiator  -- publishes <prefix>/request, subscribes <prefix>/response
  responder  -- publishes <prefix>/response, subscribes <prefix>/request

Usage:
  python host_pubsub.py                          # initiator (default)
  python host_pubsub.py --role responder
  python host_pubsub.py --prefix espp/rtps_example --domain 0

Requirements:
  pip install fastdds          # eProsima Fast DDS Python bindings
"""

from __future__ import annotations

import argparse
import signal
import time

try:
    import fastdds
except ImportError as exc:
    raise SystemExit(
        "Fast DDS Python bindings not found.\n"
        "Install the package that provides the fastdds module and retry."
    ) from exc

# ---------------------------------------------------------------------------
# Custom DDS type: raw null-terminated string, name matches the firmware.
# embeddedRTPS puts the user-supplied bytes directly into the DATA submessage
# payload with no CDR encapsulation prefix, so we do the same here.
# ---------------------------------------------------------------------------

_TYPE_NAME = "std_msgs::msg::String"


class RawStringType(fastdds.TopicDataType):
    def __init__(self) -> None:
        super().__init__()
        self.set_name(_TYPE_NAME)

    def get_name(self) -> str:
        return _TYPE_NAME

    def create_data(self) -> bytearray:
        return bytearray()

    def delete_data(self, data) -> None:
        return None

    def is_bounded(self) -> bool:
        return False

    def is_plain(self, data_representation) -> bool:
        return False

    def get_max_serialized_size_ctx(self, context) -> int:
        return 0

    def calculate_serialized_size(self, data, data_representation) -> int:
        return len(data) + 1  # +1 for null terminator

    def serialize(self, data: bytearray, payload, data_representation) -> bool:
        """Write raw bytes followed by a null terminator — matches embedded side."""
        if isinstance(data, str):
            data = data.encode()
        raw = bytes(data).rstrip(b"\x00") + b"\x00"
        return self._write_payload(payload, raw)

    def deserialize(self, payload, data: bytearray) -> bool:
        """Read raw bytes and strip the null terminator."""
        raw = self._read_payload(payload)
        text = raw.rstrip(b"\x00")
        data.clear()
        data.extend(text)
        return True

    # --- helpers -----------------------------------------------------------

    @staticmethod
    def _write_payload(payload, raw: bytes) -> bool:
        if hasattr(payload, "data"):
            try:
                payload.data[: len(raw)] = raw
            except Exception:
                payload.data = bytearray(raw)
        else:
            payload.data = bytearray(raw)
        if hasattr(payload, "length"):
            payload.length = len(raw)
        return True

    @staticmethod
    def _read_payload(payload) -> bytes:
        length = getattr(payload, "length", 0)
        buf = getattr(payload, "data", None)
        if buf is None:
            return b""
        try:
            return bytes(buf[:length])
        except Exception:
            return bytes(buf)


# ---------------------------------------------------------------------------
# Listener that prints received messages
# ---------------------------------------------------------------------------

class StringListener(fastdds.DataReaderListener):
    def __init__(self, label: str) -> None:
        super().__init__()
        self._label = label
        self._type = RawStringType()

    def on_data_available(self, reader) -> None:
        data = self._type.create_data()
        info = fastdds.SampleInfo()
        ret = reader.take_next_sample(data, info)
        ok = getattr(fastdds, "RETCODE_OK", 0)
        if ret == ok and info.valid_data:
            text = bytes(data).decode(errors="replace")
            print(f"[rx {self._label}] {text}")


# ---------------------------------------------------------------------------
# DDS participant / topic / reader / writer helpers
# ---------------------------------------------------------------------------

RETCODE_OK = getattr(fastdds, "RETCODE_OK", 0)


def _create_participant(domain_id: int, interface_ip: str | None = None) -> fastdds.DomainParticipant:
    factory = fastdds.DomainParticipantFactory.get_instance()
    qos = factory.get_default_participant_qos()

    if interface_ip:
        transport = fastdds.UDPv4TransportDescriptor()
        transport.interfaceWhiteList = [interface_ip]
        transport.sendBufferSize = 65536
        transport.receiveBufferSize = 65536
        qos.transport().use_builtin_transports = False
        qos.transport().user_transports = [transport]

    # Explicitly advertise SPDP discovery on the RTPS multicast group.
    peer = fastdds.Locator()
    peer.kind = fastdds.LOCATOR_KIND_UDPv4
    peer.address = "239.255.0.1"
    peer.port = 7400
    qos.wire_protocol().builtin.initialPeersList = [peer]

    participant = factory.create_participant(domain_id, qos)
    if participant is None:
        raise RuntimeError("Failed to create DDS participant")
    return participant


def _register_type(participant, data_type: RawStringType) -> None:
    ret = participant.register_type(data_type)
    if ret != RETCODE_OK:
        raise RuntimeError(f"Failed to register type: {ret}")


def _create_topic(participant, topic_name: str, type_name: str):
    qos = participant.get_default_topic_qos()
    topic = participant.create_topic(topic_name, type_name, qos)
    if topic is None:
        raise RuntimeError(f"Failed to create topic '{topic_name}'")
    return topic


def _create_writer(participant, topic):
    pub_qos = participant.get_default_publisher_qos()
    publisher = participant.create_publisher(pub_qos)
    if publisher is None:
        raise RuntimeError("Failed to create publisher")

    dw_qos = publisher.get_default_datawriter_qos()
    writer = publisher.create_datawriter(topic, dw_qos)
    if writer is None:
        raise RuntimeError("Failed to create data writer")
    return publisher, writer


def _create_reader(participant, topic, listener):
    sub_qos = participant.get_default_subscriber_qos()
    subscriber = participant.create_subscriber(sub_qos)
    if subscriber is None:
        raise RuntimeError("Failed to create subscriber")

    dr_qos = subscriber.get_default_datareader_qos()
    reader = subscriber.create_datareader(topic, dr_qos, listener)
    if reader is None:
        raise RuntimeError("Failed to create data reader")
    return subscriber, reader


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def run(role: str, prefix: str, domain_id: int, period_ms: int, interface_ip: str | None = None) -> None:
    if role == "initiator":
        pub_topic_name = f"{prefix}/request"
        sub_topic_name = f"{prefix}/response"
    else:
        pub_topic_name = f"{prefix}/response"
        sub_topic_name = f"{prefix}/request"

    print(f"Role      : {role}")
    print(f"Domain    : {domain_id}")
    print(f"Publish   : {pub_topic_name}")
    print(f"Subscribe : {sub_topic_name}")
    print()

    participant = _create_participant(domain_id, interface_ip)
    data_type = RawStringType()
    _register_type(participant, data_type)

    pub_topic = _create_topic(participant, pub_topic_name, _TYPE_NAME)
    sub_topic = _create_topic(participant, sub_topic_name, _TYPE_NAME)

    listener = StringListener(sub_topic_name)
    publisher, writer = _create_writer(participant, pub_topic)
    subscriber, reader = _create_reader(participant, sub_topic, listener)

    # Allow discovery to complete before the first write.
    print("Waiting for discovery (2 s)...")
    time.sleep(2.0)

    stop = False

    def _sigint(_sig, _frame):
        nonlocal stop
        stop = True

    signal.signal(signal.SIGINT, _sigint)

    counter = 0
    print("Running — Ctrl-C to stop\n")
    try:
        while not stop:
            if role == "initiator":
                text = f"request {counter}"
            else:
                text = f"response {counter}"

            sample = bytearray(text.encode())
            ret = writer.write(sample)
            if ret == RETCODE_OK:
                print(f"[tx {pub_topic_name}] {text}")
            else:
                print(f"[tx ERROR {ret}] {text}")

            counter += 1
            time.sleep(period_ms / 1000.0)
    finally:
        participant.delete_contained_entities()
        fastdds.DomainParticipantFactory.get_instance().delete_participant(participant)
        print("\nShutdown complete.")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="PC pub/sub peer for the rtps_embedded ESP32 example"
    )
    parser.add_argument(
        "--role",
        choices=["initiator", "responder"],
        default="initiator",
        help="Mirror the firmware role (default: initiator)",
    )
    parser.add_argument(
        "--prefix",
        default="espp/rtps_example",
        help="Topic prefix used in the firmware (default: espp/rtps_example)",
    )
    parser.add_argument(
        "--domain",
        type=int,
        default=0,
        help="RTPS domain ID, must match CONFIG_RTPS_EXAMPLE_DOMAIN_ID (default: 0)",
    )
    parser.add_argument(
        "--period-ms",
        type=int,
        default=2000,
        help="Publish interval in milliseconds (default: 2000)",
    )
    parser.add_argument(
        "--interface-ip",
        default=None,
        help="Optional local interface IP to constrain Fast DDS to (for example 192.168.4.2)",
    )
    args = parser.parse_args()
    run(args.role, args.prefix, args.domain, args.period_ms, args.interface_ip)


if __name__ == "__main__":
    main()

File: components/rtps/example/pc/host_pubsub.cpp
#include <atomic>
#include <chrono>
#include <csignal>
#include <cstring>
#include <iostream>
#include <memory>
#include <string>
#include <thread>

#include <fastdds/dds/domain/DomainParticipant.hpp>
#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
#include <fastdds/dds/publisher/DataWriter.hpp>
#include <fastdds/dds/publisher/Publisher.hpp>
#include <fastdds/dds/publisher/qos/DataWriterQos.hpp>
#include <fastdds/dds/subscriber/DataReader.hpp>
#include <fastdds/dds/subscriber/DataReaderListener.hpp>
#include <fastdds/dds/subscriber/SampleInfo.hpp>
#include <fastdds/dds/subscriber/Subscriber.hpp>
#include <fastdds/dds/subscriber/qos/DataReaderQos.hpp>
#include <fastdds/dds/topic/Topic.hpp>
#include <fastdds/dds/topic/TypeSupport.hpp>
#include <fastdds/rtps/common/Locator.hpp>
#include <fastdds/rtps/transport/UDPv4TransportDescriptor.hpp>

namespace {

using namespace eprosima::fastdds::dds;

static constexpr uint32_t kDomainId = 0;
static constexpr uint16_t kSpdpMulticastPort = 7400;
static constexpr char kTypeName[] = "std_msgs::msg::String";

class RawStringPubSubType : public TopicDataType {
public:
  RawStringPubSubType() {
    set_name(kTypeName);
    max_serialized_type_size = 256;
    is_compute_key_provided = false;
  }

  ~RawStringPubSubType() override = default;

  bool serialize(const void *const data, eprosima::fastdds::rtps::SerializedPayload_t &payload,
                 DataRepresentationId_t /*rep*/) override {
    const std::string *str = static_cast<const std::string *>(data);
    const uint32_t slen = static_cast<uint32_t>(str->size()) + 1u;
    const uint32_t needed = 4u + 4u + slen; // encap header + CDR length prefix + string + null
    if (needed > payload.max_size) {
      return false;
    }
    // CDR_LE encapsulation header
    payload.data[0] = 0x00;
    payload.data[1] = 0x01;
    payload.data[2] = 0x00;
    payload.data[3] = 0x00;
    // CDR string: 4-byte LE length followed by string bytes and null terminator
    payload.data[4] = static_cast<uint8_t>(slen & 0xFF);
    payload.data[5] = static_cast<uint8_t>((slen >> 8) & 0xFF);
    payload.data[6] = static_cast<uint8_t>((slen >> 16) & 0xFF);
    payload.data[7] = static_cast<uint8_t>((slen >> 24) & 0xFF);
    std::memcpy(payload.data + 8, str->c_str(), slen);
    payload.length = needed;
    payload.encapsulation = CDR_LE;
    return true;
  }

  bool deserialize(eprosima::fastdds::rtps::SerializedPayload_t &payload, void *data) override {
    // payload.data includes the 4-byte CDR encapsulation header followed by the CDR string body
    if (payload.length < 9u) {
      return false;
    }
    auto *str = static_cast<std::string *>(data);
    // Skip encapsulation header, read CDR 4-byte LE string length
    const uint32_t slen = static_cast<uint32_t>(payload.data[4]) |
                          (static_cast<uint32_t>(payload.data[5]) << 8) |
                          (static_cast<uint32_t>(payload.data[6]) << 16) |
                          (static_cast<uint32_t>(payload.data[7]) << 24);
    if (slen == 0 || 8u + slen > payload.length) {
      return false;
    }
    str->assign(reinterpret_cast<const char *>(payload.data + 8), slen - 1);
    return true;
  }

  uint32_t calculate_serialized_size(const void *const data,
                                     DataRepresentationId_t /*rep*/) override {
    const std::string *str = static_cast<const std::string *>(data);
    return 4u + 4u + static_cast<uint32_t>(str->size()) + 1u;
  }

  bool compute_key(eprosima::fastdds::rtps::SerializedPayload_t & /*payload*/,
                   eprosima::fastdds::rtps::InstanceHandle_t & /*handle*/,
                   bool /*force_md5*/) override {
    return false;
  }

  bool compute_key(const void *const /*data*/,
                   eprosima::fastdds::rtps::InstanceHandle_t & /*handle*/,
                   bool /*force_md5*/) override {
    return false;
  }

  void *create_data() override { return new std::string(); }

  void delete_data(void *data) override { delete static_cast<std::string *>(data); }

  void register_type_object_representation() override {}

#ifdef TOPIC_DATA_TYPE_API_HAS_IS_BOUNDED
  bool is_bounded() const override { return false; }
#endif
#ifdef TOPIC_DATA_TYPE_API_HAS_IS_PLAIN
  bool is_plain(DataRepresentationId_t) const override { return false; }
#endif
};

class StringListener : public DataReaderListener {
public:
  explicit StringListener(std::string label)
      : label_(std::move(label)) {}

  void on_data_available(DataReader *reader) override {
    std::string sample;
    SampleInfo info;
    while (reader->take_next_sample(&sample, &info) == RETCODE_OK) {
      if (info.valid_data) {
        std::cout << "[rx " << label_ << "] " << sample << std::endl;
      }
    }
  }

private:
  std::string label_;
};

std::atomic<bool> g_running{true};

void signal_handler(int) { g_running.store(false); }

} // namespace

int main(int argc, char **argv) {
  std::signal(SIGINT, signal_handler);
  std::signal(SIGTERM, signal_handler);

  std::string interface_ip = "192.168.4.2";
  int period_ms = 2000;

  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if (arg == "--interface-ip" && i + 1 < argc) {
      interface_ip = argv[++i];
    } else if (arg == "--period-ms" && i + 1 < argc) {
      period_ms = std::stoi(argv[++i]);
    } else {
      std::cerr << "Usage: " << argv[0] << " [--interface-ip <ip>] [--period-ms <ms>]" << std::endl;
      return 1;
    }
  }

  constexpr const char *pub_topic = "pc_to_mcu";
  constexpr const char *sub_topic = "mcu_to_pc";
  std::cout << "Publish: " << pub_topic << "\nSubscribe: " << sub_topic << std::endl;

  auto *factory = DomainParticipantFactory::get_instance();
  DomainParticipantQos qos;
  factory->get_default_participant_qos(qos);

  if (!interface_ip.empty()) {
    auto transport = std::make_shared<eprosima::fastdds::rtps::UDPv4TransportDescriptor>();
    transport->interfaceWhiteList.push_back(interface_ip);
    transport->sendBufferSize = 65536;
    transport->receiveBufferSize = 65536;
    qos.transport().use_builtin_transports = false;
    qos.transport().user_transports.push_back(transport);
  }

  eprosima::fastdds::rtps::Locator_t peer_mcast;
  peer_mcast.kind = LOCATOR_KIND_UDPv4;
  eprosima::fastdds::rtps::IPLocator::setIPv4(peer_mcast, "239.255.0.1");
  peer_mcast.port = kSpdpMulticastPort;
  qos.wire_protocol().builtin.initialPeersList.push_back(peer_mcast);

  auto *participant = factory->create_participant(kDomainId, qos);
  if (participant == nullptr) {
    std::cerr << "Failed to create participant" << std::endl;
    return 1;
  }

  TypeSupport type_support(new RawStringPubSubType());
  if (type_support.register_type(participant) != RETCODE_OK) {
    std::cerr << "Failed to register type" << std::endl;
    factory->delete_participant(participant);
    return 1;
  }

  auto *topic_pub =
      participant->create_topic(pub_topic, type_support.get_type_name(), TOPIC_QOS_DEFAULT);
  auto *topic_sub =
      participant->create_topic(sub_topic, type_support.get_type_name(), TOPIC_QOS_DEFAULT);
  if (topic_pub == nullptr || topic_sub == nullptr) {
    std::cerr << "Failed to create topics" << std::endl;
    factory->delete_participant(participant);
    return 1;
  }

  auto *publisher = participant->create_publisher(PUBLISHER_QOS_DEFAULT);
  auto *subscriber = participant->create_subscriber(SUBSCRIBER_QOS_DEFAULT);
  if (publisher == nullptr || subscriber == nullptr) {
    std::cerr << "Failed to create publisher/subscriber" << std::endl;
    participant->delete_topic(topic_pub);
    participant->delete_topic(topic_sub);
    factory->delete_participant(participant);
    return 1;
  }

  StringListener listener(sub_topic);
  auto *reader = subscriber->create_datareader(topic_sub, DATAREADER_QOS_DEFAULT, &listener);
  auto *writer = publisher->create_datawriter(topic_pub, DATAWRITER_QOS_DEFAULT);
  if (reader == nullptr || writer == nullptr) {
    std::cerr << "Failed to create reader/writer" << std::endl;
    participant->delete_subscriber(subscriber);
    participant->delete_publisher(publisher);
    participant->delete_topic(topic_pub);
    participant->delete_topic(topic_sub);
    factory->delete_participant(participant);
    return 1;
  }

  std::this_thread::sleep_for(std::chrono::seconds(2));

  uint32_t counter = 0;
  while (g_running.load()) {
    std::string msg = "pc " + std::to_string(counter++);
    if (writer->write(&msg) == RETCODE_OK) {
      std::cout << "[tx] " << msg << std::endl;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(period_ms));
  }

  subscriber->delete_datareader(reader);
  publisher->delete_datawriter(writer);
  participant->delete_subscriber(subscriber);
  participant->delete_publisher(publisher);
  participant->delete_topic(topic_pub);
  participant->delete_topic(topic_sub);
  factory->delete_participant(participant);
  return 0;
}
